File: leetcode/p0438/Solution.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int[] pMap = new int[26];
        int pLen = p.length();
        for(char c : p.toCharArray()) {
            pMap[c - 'a']++;
        }
        
        List<Integer> ans = new ArrayList<>();
        int[] sMap = new int[26];
        int start = 0;
        for (int i = 0; i < s.length(); ++i) {
            if (pLen < i - start + 1) {
                sMap[s.charAt(start) - 'a']--;
                start++;
            }
            sMap[s.charAt(i) - 'a']++;
            if (pLen == i - start + 1 && Arrays.equals(pMap, sMap)) {
                ans.add(start);
            }
        }
        return ans;
    }
}

File: leetcode/p0438/Solution.kt
class Solution {
    fun findAnagrams(s: String, p: String): List<Int> {
        val pMap = IntArray(26) {0}
        for (c in p) {
            pMap[c - 'a']++
        }
        //println(pMap.toList())
        val sMap = IntArray(26) {0}
        var start = 0
        val ans = mutableListOf<Int>()
        for(i in 0 until s.length) {
            if (p.length < (i - start + 1)) {
                sMap[s[start] - 'a']--
                start++
            }
            sMap[s[i] - 'a']++
            //println(sMap.toList())
            if (p.length == (i - start + 1) && pMap contentEquals sMap) {
                ans.add(start)
            }
        }
        return ans
    }
}
